service/src/java/org/apache/hive/service/cli/thrift/ThriftBinaryCLIService.java

```diff
@@ -113,43 +113,68 @@ protected void initServer() {
-            }
+            metrics.decrementCounter(MetricsConstant.OPEN_CONNECTIONS);
           }
-          ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
-          SessionHandle sessionHandle = context.getSessionHandle();
-          if (sessionHandle != null) {
-            LOG.info("Session disconnected without closing properly. ");
+
+          final ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
+          final Optional<SessionHandle> sessionHandle = context.getSessionHandle();
+
+          if (sessionHandle.isPresent()) {
+            // Normally, the client should politely inform the server it is
+            // closing its session with Hive before closing its network
+            // connection. However, if the client connection dies for any reason
+            // (load-balancer round-robin configuration, firewall kills
+            // long-running sessions, bad client, failed client, timed-out
+            // client, etc.) then the server will close the connection without
+            // having properly cleaned up the Hive session (resources,
+            // configuration, logging etc.). That needs to be cleaned up now.
+            LOG.warn(
+                "Client connection bound to {} unexpectedly closed: closing this Hive session to release its resources. "
+                    + "The connection processed {} total messages during its lifetime of {}ms. Inspect the client connection "
+                    + "for time-out, firewall killing the connection, invalid load balancer configuration, etc.",
+                sessionHandle, context.getMessagesProcessedCount(), context.getDuration().toMillis());
             try {
-              boolean close = cliService.getSessionManager().getSession(sessionHandle).getHiveConf()
+              final boolean close = cliService.getSessionManager().getSession(sessionHandle.get()).getHiveConf()
                   .getBoolVar(ConfVars.HIVE_SERVER2_CLOSE_SESSION_ON_DISCONNECT);
-              LOG.info((close ? "" : "Not ") + "Closing the session: " + sessionHandle);
               if (close) {
-                cliService.closeSession(sessionHandle);
+                cliService.closeSession(sessionHandle.get());
+              } else {
+                LOG.warn("Session not actually closed because configuration {} is set to false",
+                    ConfVars.HIVE_SERVER2_CLOSE_SESSION_ON_DISCONNECT.varname);
               }
             } catch (HiveSQLException e) {
-              LOG.warn("Failed to close session: " + e, e);
+              LOG.warn("Failed to close session", e);
+            }
+          } else {
+            // There is no session handle because the client gracefully closed
+            // the session *or* because the client had some issue and never was
+            // able to create one in the first place
+            if (!context.getSessionHandle().isPresent()) {
```

Comment:
nit: ```context.getSessionHandle().isPresent()``` may always return false here when executing the  ```else``` branch...

For connection exiting with ```CloseSession```, the ```sessionHandle``` is cleared before the thrift server calling the ``deleteContext```,  the logging here may be something confusing...

Reply:
Yes. Excellent. Thanks.

Reply:
Instead of keeping the last 4 session handles, I am simply keeping a counter of the total number of session handles (though, in practice, this counter will only have a value of zero or one).

Reply:
@dengzhhu653 Please review :)

Reply:
Ugh. This is still not quite correct.  It tries to print the session even though the session is non-existent.

Reply:
Removed it with latest change set.  Please review.